### PR TITLE
[WIP] Refacto Neural Model

### PR DESCRIPTION
**Pull request by hugo-quantmetry**
# Description
Warning : 
This Merge Request should not be merged now. 
It is created to start a discussion about the code philosophy in Melusine.

The NeuralModel class is outdated and need refactoring. Exemples of pain points in the current version are:
- Model saving does not work well (requires a raw pickle dump of the NeuralModel class)
- The model object used in inference contains training parameters (ex: epochs, loss, etc)
- Most NeuralModel methods are forked (if/elif) to apply different processing for CNN and Transformer models
- The NeuralModel class only be customized in terms of model architecture but not in terms of data preparation
- The model training config is hard to find (lost in long conf file)

# Suggestions
- Object Oriented Programming : Create a BaseMelusineModel class that defines a signature for all Melusine models
- Create subclasses for the CNN and Transformers models
- Make it easy for a Melusine user to customize the default models (model architecture and preprocessing)
- Create a Trainer class to separate training from inference parameters
- Improve the model saving to make it robust and make configs readable by humans
- Make the tokenizer a part of the model object

Fixes # (issue)
Several Issues are concerned:
- #102 Tokenizer
- Others (to be created)

## Type of change

Please delete options that are not relevant.

- [ x] General refactoring

# How Has This Been Tested?

The refactored code should come with an extensive set of new tests

### CODE REVIEW

File: melusine/models/models_v2/default_nn_architectures.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    cnn_net = Dense(250, activation="linear")(cnn_net)
+    cnn_net = LeakyReLU(alpha=0.05)(cnn_net)
+    cnn_net = Dense(150, activation="linear")(cnn_net)
+    cnn_net = Dropout(0.15)(cnn_net)
+    cnn_net = LeakyReLU(alpha=0.05)(cnn_net)
+
+    return cnn_net
+
+
+def default_multiclass_archi(input_net, n_targets):
```
Comment (benoitLebreton-perso) [inline melusine/models/models_v2/default_nn_architectures.py:57]: love this part because it shall help with the improvment I would like to do in https://github.com/MAIF/melusine/issues/104. Indeed I could just replace the output part or just the CNN part. I would allow us to change just the parts that need changes without any redoundencies in the code

File: melusine/models/models_v2/cnn_model.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+from melusine.models.models_v2.default_nn_architectures import (
+    default_cnn_archi,
+    default_meta_archi,
+    default_dense_archi,
+    default_multiclass_archi,
+)
+
+logger = logging.getLogger(__name__)
+
+
```
Comment (benoitLebreton-perso) [inline melusine/models/models_v2/cnn_model.py:21]: all the child classes that inherit `BaseMelusineModel` shall replace `NeuralModel` right ?
And you think about a  `CnnMelusineModel`,  `RnnMelusineModel`, ...
Good to separate BertModels and CNN in my opinion because it will allow you to delete all the `if self.architecture_function.__name__ != "bert_model"` (it occurs 5 times in NeuralModel, I don't like it).

But I think CnnModel and RnnModel should not be separated so far because they have the same methods (the same tokenizer, same metadata, same transform methods, ...)


Or have another step of inheritance : 
`class RNNorCNNMelusineModel(BaseMelusineModel)`
(help me find a better name please ^^)
with the common transform method, the common tokenizer, ...
and two child classes : 
`RnnMelusineModel(RNNorCNNMelusineModel)`
and 
`CnnMelusineModel(RNNorCNNMelusineModel)`

File: melusine/models/models_v2/base_model.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    def transform(self, x: pd.DataFrame) -> List[np.ndarray]:
+        pass
+
+    def encode_labels(self, y: pd.Series) -> np.ndarray:
+        return to_categorical(self.label_encoder.transform(y))
+
+    def save_model(self, path: str) -> None:
+        self.model.save(path)
+        logger.info(f"Saved Deep Learning model to {path}")
+
```
Comment (benoitLebreton-perso) [inline melusine/models/models_v2/base_model.py:147]: how about a MLFLOW Connector ? With the equivalent of [autolog](https://www.mlflow.org/docs/latest/python_api/mlflow.sklearn.html) method for sklearn ?
Maybe too soon and out of scope for now but maybe not so complicated

File: melusine/models/models_v2/cnn_model.py (hunk 2 above)
Comment (hugo-quantmetry) [reply melusine/models/models_v2/cnn_model.py:21]: Good point, thanks
Actually I think you can already define an RNN model using the Current CnnMelusineModel
by just changing the cnn_archi attribute.

We could try the following:
- Change the class name to be generic (WordEmbeddingBasedMelusineModel? EmbeddingBasedModel? LegacyNlpModel? HunggingfaceIsWayTooSimpleMelusineModel? )
- Add an RNN archi in the default architectures
- Describe how to use an RNN in a tutorial

File: melusine/models/models_v2/base_model.py (hunk 3 above)
Comment (hugo-quantmetry) [reply melusine/models/models_v2/base_model.py:147]: I think/hope this could be achieved through the MelusineTrainer class.
PyTorch / HuggingFace trainers make it easy to link model training to MlFlow, I'll have a look at it.

On top of that I think the Trainer should implement EarlyStopping, Logging and Tensorboard callbacks.
(Will give it a try soon)
